### DISCUSSION TITLE
Partials in multi-page apps

### DISCUSSION THREAD

**Discussion by mauf3r** [Avaiga/taipy#1527] [Q&A] [answered]
Hi there,

I am new to Taipy, and doing my first steps building a multi-page app, which should show some data which are getting updated every once in a while.
I understood the basics, my question is: can I call the app_partial method from inside a page class - and if so, how?

my basic code:
In main.py:

```
pages = {
    'Page1': Page1(),
    'Page2': Page2(),
}
Gui(pages=pages).run()
```

in Page1.py:
```
class Page1(Page):
    def __init__(self):
        super().__init__()

    def page1(self):
        with tgb.part("container"):
            tgb.text(f"# Hello World {self.username}", mode='md')

    def create_page(self):
        with tgb.Page() as page:
            self.page1()
        return page
```
        
I would like to be able to initiate/add the partial in the Page1 class. But as far as I understand I have to do it on Gui level, like:
```
   gui = Gui()
   prompt_user = gui.add_partial("")
   gui.run()
```
But If I have to place the code in main.py, I would need  to change main.py whenever I want to add a partial in a sub-page.

Any hint would be appreciated

**Comment by FlorianJacta** [comment]
I don't know where you got this syntax (which is a great syntax), however, there has yet to be any documentation on it.

You can follow [this article](https://docs.taipy.io/en/develop/tutorials/visuals/4_partials/#__tabbed_1_2) to understand how partials work.

Here is basically how you would use this partial on page 1. Partials are global variables, but they can be put on a subpage.

```python
class Page1(Page):
    def __init__(self):
        super().__init__()

    def page1(self):
        with tgb.part("container"):
            tgb.text(f"# Hello World {self.username}", mode='md')

    def create_page(self):
        with tgb.Page() as page:
            self.page1()
            tgb.part(partial="{prompt_user}")

        return page
```

**Reply by mauf3r** [reply]
I found that approach in the demo-stock-visualization example - and I think the code is yours :-)

The article in the docs is, what I was referring to. As far as I understood, the prompt_user partial needs to be initialized in the Gui class, and I don't know if/how I can access the Gui from inside the Page class.

Here is what Chat-GPT suggested:
**main.py**
```python
from Page1 import Page1
from Page2 import Page2
from taipy.gui import Gui

# Initialisiere die Seiten
pages = {
    'Page1': Page1(),
    'Page2': Page2(),
}

# Erstelle das Gui-Objekt
gui = Gui(pages=pages)

# Füge Partials aus den Seitenklassen hinzu
for page in pages.values():
    if hasattr(page, add_partials)
        page.page.add_partials(gui)

# Starte die Anwendung
gui.run()
```

**Page1.py**
```python
from taipy.gui import Page
import taipy.gui as tgb

class Page1(Page):
    def __init__(self):
        super().__init__()
        self.username = "User"

    def update_page1_partial(self):
        with tgb.Page as part_update:
            # do some stuff
            pass
        self.page1_partial.update_contrent(self, part_update)

    def page1(self):
        with tgb.part("container"):
            tgb.text(f"# Hello World {self.username}", mode='md')
            tgb.part(partial="{page1_partial}")

    def create_page(self):
        with tgb.Page() as page:
            self.page1()
        return page

    def add_partials(self, gui):
        gui.add_partial("page1_partial", self.updtate_page1_partal)
```

**Reply by FlorianJacta** [reply]
Yes, this is my code 😄  I suggest sticking to the main method of creating Taipy applications until the documentation is ready, or you will encounter other issues.

```python
from Page1 import page_1, update_page1_partial
from Page2 import page_2
from taipy.gui import Gui

# Initialisiere die Seiten
pages = {
    'Page1': page_1,
    'Page2': page_2
}

def on_init(state):
    update_page1_partial(state)

# Erstelle das Gui-Objekt
gui = Gui(pages=pages)

page1_partial= gui.add_partial("")

# Starte die Anwendung
gui.run()
```

Page1.py


```python
import taipy.gui.builder as tgb

username = "User"

def update_page1_partial(state):
    with tgb.Page() as part_update:
        # do some stuff
        tgb.text('Do some stuff')
     state.page1_partial.update_content(state, part_update)

def page1():
    with tgb.part("container"):
        tgb.text("# Hello World {username}", mode='md')
        tgb.part(partial="{page1_partial}")

with tgb.Page() as page_1:
    page1()
```

I hope this helps!

**Reply by mauf3r** [reply]
I appreciate your advice, but I like this approach very much. So I will follow that path, you showed in your example.

I assume, that most of the issues I am going to have will be more basic. So I hope It is OK, if I continue to ask questions ☺️

**Reply by FlorianJacta** [reply, marked as answer]
Yes, it is ok. I am just unsure if I will be able to answer you as it has not been thoroughly tested
